doc/repository-spec-v2.md

```diff
@@ -1,39 +1,216 @@
+
+The `User-Agent` header also allows package repository to determine how many
+different clients would be affected by an API change.
+
+
+## Retrying with Exponential Backoff
+The `dart pub` client will retry failed requests with
+[exponential backoff](https://en.wikipedia.org/wiki/Exponential_backoff).
+This aims to increase robustness against intermittent network issues, while not
+overloading servers that are partially failing.
+
+Clients are strongly encouraged to employ exponential backoff starting at 200ms,
+400ms, etc. stopping after 5-7 retries. Excessive can have negative impact on
+servers and network performance.
+
+
+## Rejecting Requests
+The `dart pub` client will in many cases to display error messages when given a
+response as follows:
+
+```http
+HTTP/1.1 4XX OK
+Content-Type: application/vnd.pub.v2+json
+{
+  "error": {
+    "code": "<code>",
+    "message": "<message>",
+  },
+}
+```
+
+The `<message>` is intended to be a brief human readable explanation of what
+when wrong and why the request failed. The `<code>` is a text string intended to
+allow clients to handle special cases without using regular expression to
+parse the `<message>`.
+
+
+## Authentication
+The `dart pub` client allows users to save an opaque `<token>` for each
+`<hosted-url>`. When the `dart pub` client makes a request to a `<hosted-url>`
+for which it has a `<token>` stored, it will attach an `Authorization` header
+as follows: 
+
+ * `Authorization: Bearer <token>`
+
+Tokens can be added to `dart pub` client using the command:
+
+ *  `dart pub token add <hosted-url>`.
+
+This command will prompt the user for the `<token>` on stdin, reducing the risk
+that the `<token>` is accidentally stored in shell history. For further details
+on token management see: `dart pub token --help`.
+
+
+### Missing Authentication or Invalid Token
+If the server requires authentication and the request does not carry an
+`Authorization` header, or the token in the `Authorization` header is invalid
+or expired, the server must respond:
+
+```http
+HTTP/1.1 401 Unauthorized
+WWW-Authenticate: Bearer realm="pub", message="<message>"
+```
+
+If the `dart pub` client receives a `401` response and the `dart pub` client has
+a token for the given `<hosted-url>`, then the `dart pub` client knows for sure
+that the token it has stored for the given `<hosted-url>` is invalid.
+Hence, the `dart pub` client shall remove the token from local configuration.
+Hence, a server shall not send `401` in case where a token is valid, but does
+not have permissions to access the package in question.
+
+When receiving a `401` response the `dart pub` client shall:
+ * Abort the current operation (exiting non-zero),
+ * Delete any token it may have stored for the given `<hosted-url>`,
+ * Inform the user that authentication is required,
+ * Print the `<message>` provided by the server.
+
+The `<message>` allows a custom _package server_ to how a token may be obtained.
```

Comment:
The sentence seems incomplete. I might be wrong, but I think it needs ".. to inform user how a token..."